src/julia.jl

```
@@ -33,4 +33,56 @@ function multireshape(parent::AbstractArray, dimss::Array{Tuple{Vararg{Int}}})
     multireshape(parent, dimss...)
 end
 
+"""
+    pairwise(f::Function, a; kwargs...)
+
+Apply the function `f` to all pairwise combinations of elements from `a`. Set
+keyword `symmetric = true` to only calculate `f(x,y)` when `f(x,y) == f(y,x)`.
+"""
+function pairwise(f::Function, a; symmetric = false)
+    n = length(a)
+    first = f(a[1], a[1])
```

`firstval` = f(first(a), first(a))` would handle cases that don't use 1-based indices.  Also, you could pass `firstval` to `pairwise!` (or have an internal `_pairwise!` function that takes an extra argument) to avoid recomputing it.

(This will still be problematic for type-unstable functions.  There is a way to fix this, though...)

---

Right, thanks! I am so curious now about the way to fix this. OTOH I'd say that it is OK for type-unstable functions to fail noisily here though?

---

Perhaps you just mean that the way to fix this is to not write type-unstable functions. Do you think it would be worth the extra code to insert the first value and avoid recomputing?

---

The way to fix this is to check the type of `f(x,y)` in the loop, and if it doesn't match then re-allocate the matrix.   In the common case of type-stable functions `f`, the check will be eliminated by the compiler and it will be fast.

This is how comprehensions, `map`, and `broadcast` do it.

---

i.e. you do something like:
```
result = f(a[i],a[j])
if isa(result, eltype(r))
    r[i,j] = result
else
    rnew = similar(r, Union{eltype(r),typeof(result)})
    ... copy r to rnew for elements already written ...
    rnew[i,j] = result
    return pairwise_symmetric!(i+1,j, f, a, rnew)
end
```
end